### PR TITLE
Unit test: module Posts

### PR DESCRIPTION
**Pull request by gareth-go**
## This pull request
- Add unit test for Post model and controller

## Checkpoint

- [ ] Checked the work against _all_ of the ticket requirements
- [ ] CHANGELOG updated
- [ ] Formatter has been run against all changed files
- [x] All changed files have been added correctly
- [x] Appropriate unit tests and integration tests have been added
- [x] All tests are passing
- [ ] Ad hoc testing has been done
- [ ] Affected docs have been updated
- [ ] All licenses have been checked and confirmed for commercial use

## Front End Specific

- [ ] All styles have been applied and responsiveness tested
- [ ] Changes have been checked on supported browsers
- [ ] Changes have been checked on supported mobile devices
- [ ] Assets are added and have been optimized
- [ ] End to end tests updated
- [ ] a11y concerns addressed
- [ ] Bundle builds and serves correctly

## Back End Specific

- [ ] Database migrations have been tested
- [ ] Necessary data patches have been prepared
- [ ] Infrastructure changes have been completed

### CODE REVIEW

File: spec/controllers/posts_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+  describe 'Show' do
+    context 'user can not see un-accepted post of other user' do
+      before { login_user }
+
+      it do
+        get :edit, params: { id: post.id }
+
+        should redirect_to root_path
+        should set_flash[:alert]
```
Comment (gary-gia-goldenowl) [inline spec/controllers/posts_controller_spec.rb]: You have similar tests for different actions (accept, reject, destroy). You can create shared examples for this to reduce duplication. for example

```
shared_examples '....' do |action|
  it do
   //
  end
end
```

File: spec/controllers/posts_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      end
+    end
+  end
+
+  describe 'Edit' do
+    context 'user can not edit post of other user' do
+      before { login_user }
+
+      it do
```
Comment (gary-gia-goldenowl) [inline spec/controllers/posts_controller_spec.rb]: Write description to each of your it blocks to clear what each test is testing, for example

it "redirect to root_path when user ..."

File: spec/controllers/posts_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
+require 'rails_helper'
+
+RSpec.describe PostsController, type: :controller do
+  let(:post) { FactoryBot.create(:post) }
+
+  describe 'Show' do
```
Comment (gary-gia-goldenowl) [inline spec/controllers/posts_controller_spec.rb]: Make sure follow[ rspec style guide](https://github.com/rubocop/rspec-style-guide) for write describe, if you write for an action it should be like

`describe 'GET #show' do`

**refactor the same for other actions**

File: spec/controllers/posts_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      it do
+        get :edit, params: { id: post.id }
+
+        should redirect_to root_path
+        should set_flash[:alert]
+      end
+    end
+  end
+
+  describe 'Edit' do
```
Comment (gary-gia-goldenowl) [inline spec/controllers/posts_controller_spec.rb]: You should write test for action 'PUT #update', 'POST #create' too not just get action

```
describe 'PUT #update' do
  let(:params) do
  end
  
  def do_request(params = {})
    put :update, params: params
  end
  
  context 'success to update ....' do
  end
  
  context 'fail to update....' do
  end
end
```

File: spec/factories/user.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+FactoryBot.define do
+  factory :user do
+    sequence(:username) { |n| "user-test-#{n}" }
+    sequence(:email) { |n| "test-#{n}@sample.com" }
+    password { '123456' }
+
+    factory :admin do
+      after(:create) { |user| user.update(role: :admin) }
+    end
```
Comment (gary-gia-goldenowl) [inline spec/factories/user.rb]: You should use **trait** instead

File: spec/models/post_spec.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    it { should have_many(:reactions).dependent(:destroy) }
+    it { should have_many(:comments).dependent(:destroy) }
+    it { should have_many(:notifications).dependent(:destroy) }
+  end
+
+  describe 'validations' do
+    it { should define_enum_for(:status).with_values(%i[pending accepted rejected]) }
+    it { should validate_presence_of(:title).with_message('Title can not be blank') }
+    it { should validate_uniqueness_of(:title).with_message('This title is already exist') }
+  end
```
Comment (gary-gia-goldenowl) [inline spec/models/post_spec.rb:17]: Make sure you've tested for custom validation too

Comment (gareth-go) [reply spec/models/post_spec.rb:17]: The Post model didn't have any custom validation.

File: spec/controllers/posts_controller_spec.rb [hunk 7]
```diff
@@ -1,100 +1,151 @@
...
+
+    let(:post_record) { FactoryBot.create(:post, user: controller.current_user) }
+
+    context 'with valid params' do
+      let(:new_attributes) { { title: 'new title' } }
 
-      it do
-        put :accept, params: { id: post.id }
+      it 'updates the post' do
+        put :update, params: { id: post_record.id, post: new_attributes }
```
Comment (gary-gia-goldenowl) [inline spec/controllers/posts_controller_spec.rb:74]: the `put :update, params: { id: post_record.id, post: new_attributes }` can refactored by wrap up and put after the context for example
```
describe 'PUT #update' do 
  let(:params) do
    {
    ///
    }
  end
context 'with valid params' do
  before do
    put :update, params: params
  end
  
  it 'updates the post' do
  //
  end
end

context '....'

end
```

**Pull request merged**